File: src/crewlet/secrets/cipher.py
"""Authenticated-encryption cipher for company-config secrets at rest.

A :class:`SecretCipher` seals a plaintext secret into a self-describing
``enc:v1:<key_id>:<base64>`` envelope and reverses the transform. The
concrete :class:`KeyringCipher` uses AES-256-GCM (AEAD) with a per-value
random nonce; the field's JSON-pointer path is bound in as *associated
data* so a ciphertext cannot be replayed into a different field.

The envelope carries the ``key_id`` that sealed it, so one revision can
hold ciphertexts under a mix of keys during a key rotation and each
decrypts under the right entry of the keyring.

See ``docs/concepts/configuration.md`` § Secrets.
"""

from __future__ import annotations

import base64
import binascii
import os
from typing import TYPE_CHECKING, Protocol, runtime_checkable

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

if TYPE_CHECKING:
    from crewlet.config import SecretsConfig

#: Prefix + format-version tag every envelope carries. Bumping the ``v1``
#: is how a future algorithm change stays distinguishable from existing
#: ciphertext.
ENVELOPE_PREFIX = "enc:v1:"

#: AES-256 key size in bytes.
KEY_BYTES = 32

#: GCM nonce size in bytes (96 bits — the AES-GCM standard nonce length).
NONCE_BYTES = 12


class SecretCipherError(Exception):
    """Base error for secret sealing / unsealing."""


class SecretDecryptError(SecretCipherError):
    """Raised when an envelope cannot be decrypted.

    Covers every failure mode uniformly — wrong key, unknown ``key_id``,
    tampered ciphertext, or an associated-data (field-path) mismatch —
    so a caller cannot distinguish them and use the difference as an
    oracle.
    """


@runtime_checkable
class SecretCipher(Protocol):
    """Seals and unseals a single secret string.

    Implementations must be deterministic on ``decrypt`` and bind
    ``aad`` (the field's JSON-pointer path) so a ciphertext moved to a
    different field fails to decrypt.
    """

    def encrypt(self, plaintext: str, *, aad: str) -> str:
        """Return an ``enc:v1:<key_id>:<base64>`` envelope for ``plaintext``."""
        ...

    def decrypt(self, token: str, *, aad: str) -> str:
        """Reverse :meth:`encrypt`. Raise :class:`SecretDecryptError` on failure."""
        ...


def is_envelope(value: object) -> bool:
    """True when ``value`` is a well-formed ``enc:v1:<key_id>:<blob>`` token."""
    if not isinstance(value, str) or not value.startswith(ENVELOPE_PREFIX):
        return False
    # "enc", "v1", key_id, blob — four colon-separated parts, key_id and
    # blob both non-empty.
    parts = value.split(":", 3)
    return len(parts) == 4 and bool(parts[2]) and bool(parts[3])


def parse_key_id(token: str) -> str | None:
    """Return the ``key_id`` an envelope names, or ``None`` if not an envelope."""
    if not is_envelope(token):
        return None
    return token.split(":", 3)[2]


def decode_key_material(material: str) -> bytes:
    """Decode a base64 key string into raw 32 bytes, validating length.

    ``material`` is the resolved value of a Tier A ``secrets.keys[].material``
    entry (``${VAR}`` already substituted at bootstrap load time).
    """
    try:
        raw = base64.b64decode(material, validate=True)
    except (binascii.Error, ValueError) as exc:
        raise SecretCipherError("secret key material must be valid base64") from exc
    if len(raw) != KEY_BYTES:
        raise SecretCipherError(
            f"secret key material must decode to {KEY_BYTES} bytes "
            f"(got {len(raw)}); generate one with `crewlet secrets keygen`"
        )
    return raw


class KeyringCipher:
    """AES-256-GCM cipher over a small keyring.

    ``keys`` maps ``key_id`` → raw 32-byte key; ``active_key_id`` names
    the entry used to seal new writes. Every configured key stays
    available for :meth:`decrypt`, so a rotation (new active key, old key
    retained) decrypts mixed-key revisions without downtime.
    """

    def __init__(self, keys: dict[str, bytes], active_key_id: str) -> None:
        if not keys:
            raise SecretCipherError("keyring is empty")
        if active_key_id not in keys:
            raise SecretCipherError(
                f"active_key_id {active_key_id!r} is not in the keyring"
            )
        for key_id, material in keys.items():
            if len(material) != KEY_BYTES:
                raise SecretCipherError(
                    f"key {key_id!r} is {len(material)} bytes, expected {KEY_BYTES}"
                )
        self._keys = dict(keys)
        self._active_key_id = active_key_id

    @classmethod
    def from_config(cls, config: SecretsConfig) -> KeyringCipher | None:
        """Build a cipher from Tier A ``secrets`` config, or ``None`` if unset.

        Returns ``None`` when no keys are configured — the caller treats
        that as "secret encryption disabled" and fails closed only if the
        active revision actually contains sealed values.
        """
        if not config.keys:
            return None
        keys = {k.id: decode_key_material(k.material) for k in config.keys}
        return cls(keys=keys, active_key_id=config.active_key_id)

    @property
    def active_key_id(self) -> str:
        return self._active_key_id

    def encrypt(self, plaintext: str, *, aad: str) -> str:
        key = self._keys[self._active_key_id]
        nonce = os.urandom(NONCE_BYTES)
        ct = AESGCM(key).encrypt(nonce, plaintext.encode("utf-8"), aad.encode("utf-8"))
        blob = base64.b64encode(nonce + ct).decode("ascii")
        return f"{ENVELOPE_PREFIX}{self._active_key_id}:{blob}"

    def decrypt(self, token: str, *, aad: str) -> str:
        if not is_envelope(token):
            raise SecretDecryptError("value is not an enc:v1 envelope")
        _, _, key_id, blob = token.split(":", 3)
        key = self._keys.get(key_id)
        if key is None:
            raise SecretDecryptError(
                f"no key for key_id {key_id!r} in the keyring — the sealing "
                f"key is missing from Tier A `secrets.keys`"
            )
        try:
            raw = base64.b64decode(blob, validate=True)
            nonce, ct = raw[:NONCE_BYTES], raw[NONCE_BYTES:]
            plaintext = AESGCM(key).decrypt(nonce, ct, aad.encode("utf-8"))
        except (InvalidTag, binascii.Error, ValueError) as exc:
            raise SecretDecryptError(
                "decryption failed (wrong key, tampered ciphertext, or "
                "field-path mismatch)"
            ) from exc
        return plaintext.decode("utf-8")


__all__ = [
    "ENVELOPE_PREFIX",
    "KEY_BYTES",
    "NONCE_BYTES",
    "KeyringCipher",
    "SecretCipher",
    "SecretCipherError",
    "SecretDecryptError",
    "decode_key_material",
    "is_envelope",
    "parse_key_id",
]

File: go/internal/secrets/cipher.go
// Package secrets seals configuration secrets at rest and resolves them at
// read time.
//
// Two composable mechanisms share one keyring. A Cipher seals a value into a
// self-describing envelope, used both for whole-document company-config
// encryption and for the per-variable secret store. A Source answers ${VAR}
// lookups ahead of the process environment, so a rotated secret takes effect
// without anyone editing a file.
//
// # The envelope
//
// "enc:v1:<key_id>:<base64>". It carries the id of the key that sealed it,
// so one revision can hold ciphertexts under a MIX of keys during a rotation
// and each decrypts under the right entry. The "v1" is how a future
// algorithm change stays distinguishable from existing ciphertext rather
// than silently failing to parse.
//
// # Associated data is not optional
//
// Every seal binds the value's location — a field path, or the variable's
// name in the secret store — as AEAD associated data. Without it a
// ciphertext is portable: an operator (or an attacker with write access to
// the config) could move a sealed high-privilege token into a
// lower-privilege field and the engine would decrypt it happily. With it,
// the move fails to authenticate.
package secrets

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"errors"
	"fmt"
	"strings"
)

const (
	// EnvelopePrefix tags every sealed value.
	EnvelopePrefix = "enc:v1:"

	// keyBytes is the AES-256 key size.
	keyBytes = 32
)

// ErrDecrypt is returned for EVERY unsealing failure — wrong key, unknown
// key id, tampered ciphertext, or an associated-data mismatch.
//
// Deliberately uniform: a caller that could tell those apart would have a
// decryption oracle, and the difference between "wrong key" and "wrong
// field" is exactly the signal an attacker probing a config would want.
var ErrDecrypt = errors.New("secret could not be decrypted")

// ErrNoKey means the keyring has no usable key. Distinct from ErrDecrypt
// because it is an operator configuration problem, not a data problem, and
// it must fail loudly rather than yielding an empty string — an empty secret
// becomes an empty Bearer token discovered hours later.
var ErrNoKey = errors.New("no encryption key configured")

// Cipher seals and unseals secrets.
type Cipher interface {
	// Encrypt seals plaintext, binding aad as associated data.
	Encrypt(plaintext, aad string) (string, error)
	// Decrypt unseals an envelope, requiring the same aad.
	Decrypt(token, aad string) (string, error)
}

// IsEnvelope reports whether a value is a well-formed sealed envelope.
//
// Used to tell an already-sealed value from a plaintext one so re-sealing a
// document is idempotent, and so a redaction pass can recognise ciphertext
// without attempting to decrypt it.
func IsEnvelope(value string) bool {
	_, ok := EnvelopeKeyID(value)
	return ok
}

// EnvelopeKeyID reads which keyring key sealed a value.
//
// THE ONE PARSE of the envelope's shape, because the key id is denormalised
// into a column — a rotation sweep finds stale rows by comparing it, without
// decrypting any of them — and a second reading of the format is how the
// column and the ciphertext come to disagree about which key opens a row.
//
// False for anything that is not a well-formed envelope, which is what
// [IsEnvelope] is: telling a sealed value from a plaintext one is the same
// question as "which key sealed it", answered once.
func EnvelopeKeyID(value string) (string, bool) {
	id, _, ok := splitEnvelope(value)
	return id, ok
}

// splitEnvelope is THE parse of the envelope's shape: prefix, key id,
// base64 payload. Both the key-id reader above and Decrypt below go through
// it, because two readings of one format is how a column that says which key
// opens a row comes to disagree with the ciphertext in it.
func splitEnvelope(value string) (id string, raw []byte, ok bool) {
	if !strings.HasPrefix(value, EnvelopePrefix) {
		return "", nil, false
	}
	id, blob, cut := strings.Cut(strings.TrimPrefix(value, EnvelopePrefix), ":")
	if !cut || id == "" || blob == "" {
		return "", nil, false
	}
	decoded, err := base64.StdEncoding.DecodeString(blob)
	if err != nil {
		return "", nil, false
	}
	return id, decoded, true
}

// Keyring is a set of named AES-256 keys, one of which is active.
//
// Multiple keys exist so a rotation is zero-downtime: the new key becomes
// active and seals new values, while old ciphertext keeps decrypting under
// the key that sealed it until a rekey pass rewrites it.
type Keyring struct {
	// ActiveID names the key new values are sealed under.
	ActiveID string
	// Keys maps key id to a 32-byte AES-256 key.
	Keys map[string][]byte
}

// Validate reports whether a keyring is usable.
func (k Keyring) Validate() error {
	if len(k.Keys) == 0 {
		return ErrNoKey
	}
	if k.ActiveID == "" {
		return fmt.Errorf("%w: no active key id", ErrNoKey)
	}
	if _, ok := k.Keys[k.ActiveID]; !ok {
		return fmt.Errorf("%w: active key %q is not in the keyring", ErrNoKey, k.ActiveID)
	}
	var errs []error
	for id, key := range k.Keys {
		if len(key) != keyBytes {
			errs = append(errs, fmt.Errorf("key %q is %d bytes, want %d", id, len(key), keyBytes))
		}
	}
	return errors.Join(errs...)
}

// keyringCipher is the AES-256-GCM implementation.
type keyringCipher struct {
	ring Keyring
}

// NewCipher builds a Cipher over a keyring.
func NewCipher(ring Keyring) (Cipher, error) {
	if err := ring.Validate(); err != nil {
		return nil, err
	}
	// Copy the key material so a later mutation of the caller's map
	// cannot silently change which key seals what.
	keys := make(map[string][]byte, len(ring.Keys))
	for id, k := range ring.Keys {
		keys[id] = append([]byte(nil), k...)
	}
	return &keyringCipher{ring: Keyring{ActiveID: ring.ActiveID, Keys: keys}}, nil
}

func (c *keyringCipher) aead(id string) (cipher.AEAD, error) {
	key, ok := c.ring.Keys[id]
	if !ok {
		return nil, ErrDecrypt
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, ErrDecrypt
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, ErrDecrypt
	}
	return gcm, nil
}

// Encrypt seals plaintext under the active key.
func (c *keyringCipher) Encrypt(plaintext, aad string) (string, error) {
	gcm, err := c.aead(c.ring.ActiveID)
	if err != nil {
		return "", fmt.Errorf("%w: active key unusable", ErrNoKey)
	}
	nonce := make([]byte, gcm.NonceSize())
	if _, err := rand.Read(nonce); err != nil {
		return "", fmt.Errorf("generate nonce: %w", err)
	}
	// The nonce is prefixed to the ciphertext rather than stored
	// separately: it is not secret, it must never repeat under one key,
	// and keeping it beside the bytes it belongs to removes any way for
	// the two to be separated.
	sealed := gcm.Seal(nonce, nonce, []byte(plaintext), []byte(aad))
	return EnvelopePrefix + c.ring.ActiveID + ":" + base64.StdEncoding.EncodeToString(sealed), nil
}

// Decrypt unseals an envelope under whichever key sealed it.
func (c *keyringCipher) Decrypt(token, aad string) (string, error) {
	id, raw, ok := splitEnvelope(token)
	if !ok {
		return "", ErrDecrypt
	}
	gcm, err := c.aead(id)
	if err != nil {
		return "", ErrDecrypt
	}
	if len(raw) < gcm.NonceSize() {
		return "", ErrDecrypt
	}
	nonce, ct := raw[:gcm.NonceSize()], raw[gcm.NonceSize():]
	plaintext, err := gcm.Open(nil, nonce, ct, []byte(aad))
	if err != nil {
		// Every failure mode collapses to one error here. See ErrDecrypt.
		return "", ErrDecrypt
	}
	return string(plaintext), nil
}

// GenerateKey returns a fresh random AES-256 key.
func GenerateKey() ([]byte, error) {
	key := make([]byte, keyBytes)
	if _, err := rand.Read(key); err != nil {
		return nil, fmt.Errorf("generate key: %w", err)
	}
	return key, nil
}

// EncodeKey renders a key in the form the config's `material` field takes.
//
// Base64 of the raw bytes, which is what the loader decodes — so `crewlet
// secrets keygen` emits something pasteable rather than something to
// convert, and there is one definition of the encoding rather than one here
// and one in the loader.
func EncodeKey(key []byte) string { return base64.StdEncoding.EncodeToString(key) }

// AADForVar is the associated data binding a secret-store value to its
// variable name, so a ciphertext moved between rows fails to authenticate.
func AADForVar(name string) string { return "secret_values/" + name }

// AADForDocument is the associated data for a whole-config document.
const AADForDocument = "company_config/document"
